### PR TITLE
Add support for LlamaIndex handler

### PR DESCRIPTION
**Pull request by bala-ceg**
## Description

Implemented the LlamaIndex  ML handler as per the new ML handler framework

## Type of change

(Please delete options that are not relevant)

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📢 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📄 This change requires a documentation update

### What is the solution?
- [x] LlamaIndex ML Handler
  - [x] [Support Web Page Reader](https://gpt-index.readthedocs.io/en/latest/examples/data_connectors/WebPageDemo.html)

#### Screenshot-1
Screenshot 2023-05-28 at 10 20 17 PM

#### Screenshot-2
Screenshot 2023-05-28 at 10 28 24 PM

#### Screenshot-3
Screenshot 2023-05-28 at 10 35 12 PM

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation or created issues to update them.
- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] I have shared a short loom video or screenshots demonstrating any new functionality.

### CODE REVIEW

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (Disiok) [inline mindsdb/integrations/handlers/llama_index_handler/README.md:18]: nit: is this configuration hooked up properly? In the code below I don't see this arg being used at all.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (Disiok) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:67]: Shouldn't the index class be configured to be one specified in args?

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (Disiok) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:64]: not for this PR: curious for the future, do we have plans for a deeper integration (e.g. I can imagine  we have the context data also in the input df.  We can use LlamaHub loaders to populate those fields, as well as using building indexes on top of those the context pulled out of df).

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (Disiok) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:65]: curious: how come we need to reload this upon each predict call? If I understand correctly, the predict calls only change the questions, but the context remain fixed.
I'd imagine we can build the index once and serve any number of predict calls thereafter.

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md:18]: Yes, I have used some of the args mentioned in the configuration. I have not used index_class & query_engine in the code

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:67]: I agree, I will address this change in sometime

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:65]: I agree; I will address this change in sometime

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:64]: I defer this question to @torrmal @paxcema

File: mindsdb/integrations/handlers/llama_index_handler/setup.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/setup.py:8]: The `requirements.txt` file is missing (needed as per the [expected](https://docs.mindsdb.com/contribute/ml-handlers#step-by-step-instructions) ML handler folder structure), so this would fail. 

Perhaps simply forgot to push it? Regardless, I'd recommend pinning the LlamaIndex version to avoid silent major releases breaking the integration.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:64]: Yup, I believe a deeper integration is definitely on the roadmap.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:70]: Would this support bulk predictions? 

If a MindsDB user does this:
```sql
SELECT t.question, m.answer
FROM mindsdb.my_qa_model as m
JOIN my_data.table_with_a_bunch_of_questions as t;
```

they expect as output a table with the model's answer for every question in the second clause of the join. Based on the `iat[0]`, I would imagine it will not work (maybe I'm missing something though!). 

Please double check that the above example does go through, as it should definitely be part of the initial release.

**Review by paxcema — CHANGES_REQUESTED** [review]
This looks great! One major concern is the support for bulk queries, please have a look at that when you've got some time.

**Comment by bala-ceg** [comment]
## Latest working Screenshots
Screenshot 2023-05-29 at 10 01 35 PM

Screenshot 2023-05-29 at 9 44 41 PM

Screenshot 2023-05-29 at 10 13 13 PM

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py]: Thanks, @paxcema for reviewing the PR and pointing out this.
 I fixed this and updated the README.md file with SQL usage. Please find the attached working screenshots of the same.

File: mindsdb/integrations/handlers/llama_index_handler/setup.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/setup.py:8]: My Apologies. I forgot to push the requirements.txt file.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py]: Addressed the changes. Please kindly check this

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py]: addressed the changes, please kindly check this

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/README.md:28]: Nit: if possible, add the screenshot example with `JOIN` here, too.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:77]: Is this a leftover? We shouldn't be logging errors on every predict call 👀

**Review by paxcema — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md:51]: Done. Thank you

**Comment by bala-ceg** [comment]
## Working Screenshots
Screenshot 2023-05-30 at 8 09 15 PM
Screenshot 2023-05-30 at 8 04 24 PM
Screenshot 2023-05-30 at 7 58 08 PM
Screenshot 2023-05-30 at 7 54 32 PM
Screenshot 2023-05-30 at 7 51 03 PM
Screenshot 2023-05-30 at 7 50 24 PM

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:67]: Out of ignorance (perhaps one for @Disiok), would it help to preface this doc_str with the actual column names?

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:75]: Seems like an `else: raise Exception()` is needed here. Otherwise we store `None` and it would break at prediction time.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:93]: With the suggested exception inside `create`, these two `if`s wouldn't be needed, just define `query_engine`.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:153]: Please clean this up, e.g. rename the signature parameter to `documents`

**Comment by smyja** [comment]
This looks really good @bala-ceg I'll extend it once it's merged.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:75]: Thanks for suggestion @paxcema, I have made the suggested changes

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:93]: Thanks for suggestion @paxcema, I have made the suggested changes

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:153]: Thanks for suggestion @paxcema, I have made the suggested changes

**Comment by bala-ceg** [comment]
> This looks really good @bala-ceg I'll extend it once it's merged.

Thanks @smyja. Always a pleasure

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (jerryjliu) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:67]: up to you! i think it could help

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (jerryjliu) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:154]: this is for a future PR - we actually have a `GPTPandasIndex` class that can operate directly over a pandas dataframe (and querying the index would just create a sequence of pandas operations that run in a python REPL) 

could be better than using a vector store (though may need to think about security)

**Review by jerryjliu — APPROVED** [review]
approving to unblock; in general this looks pretty good for an initial draft, my comments are pretty minor

@bala-ceg @paxcema i'm planning to present this as a slide or two in the llamaindex talk at the mindsdb conference - would the screenshots you attached be good to showcase? feel free to attach other material as well :)

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md:25]: Yes Jerry

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:84]: I will make the proposed changes

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md:41]: I thought as per the initial design doc, there were multiple query engines, I will remove this and make the proposed changes

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md:5]: I will make the proposed changes

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md:2]: I will make the proposed changes

**Comment by bala-ceg** [comment]
@jerryjliu I will add the screenshots to google drive and share here shortly

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:67]: @paxcema, Jerry wrote the following in the below thread "up to you! i think it could help"

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:56]: Thanks, Jerry for pointing out this, moved this code to inside the if ... elif ... else loop starting on line 56

File: mindsdb/integrations/handlers/llama_index_handler/README.md
Comment (bala-ceg) [reply mindsdb/integrations/handlers/llama_index_handler/README.md]: Fixed this in README.md and also in the handler file

**Comment by bala-ceg** [comment]
@jerryjliu Please find the Google drive folder link for the screenshots. Please let me know in case if you face any issues while accessing the folder.
https://drive.google.com/drive/folders/1KdS2ilwLLzTFH8_exhYvk49IswTmkHa0?usp=share_link

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (Disiok) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:79]: very nit: I found the naming of "pred" unintuitive.

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (Disiok) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:84]: Looks like this is not fixed yet?

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (Disiok) [inline mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:59]: nit: why not just have a `docs = []`

Comment (Disiok) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:59]: naming is very unintuitive.

**Review by Disiok — APPROVED** [review]
Looks good to me, just some minor nits

**Comment by Disiok** [comment]
> @jerryjliu Please find the Google drive folder link for the screenshots. Please let me know in case if you face any issues while accessing the folder. https://drive.google.com/drive/folders/1KdS2ilwLLzTFH8_exhYvk49IswTmkHa0?usp=share_link

Thanks @bala-ceg , I was able to access!

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:63]: @bala-ceg let's go for it then and initialize doc_str with `df.columns`!

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (paxcema) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:79]: AFAIK (@ea-rus can confirm) there are no limitations on naming with `file_set`, I would suggest perhaps going with `'reader'` instead.

**Review by paxcema — APPROVED** [review]
Ok, very minor points left. Let's merge and we can continue to improve in new PRs!

File: mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py
Comment (ea-rus) [reply mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py:79]: yes, no limitations

**Pull request merged**
